pkg/controller/nodekeeper/nodekeeper_controller.go

```diff
@@ -0,0 +1,195 @@
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNodeKeeper{
+		client:               mgr.GetClient(),
+		configManagerBuilder: configmanager.NewBuilder(),
+		machinery:            machinery.NewMachinery(),
+		metricsClientBuilder: metrics.NewBuilder(),
+		scheme:               mgr.GetScheme(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nodekeeper-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Node, status change will not trigger a reconcile
+	err = c.Watch(
+		&source.Kind{Type: &corev1.Node{}},
+		&handler.EnqueueRequestForObject{},
+		IgnoreMasterPredicate)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileNodeKeeper implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNodeKeeper{}
+
+// ReconcileNodeKeeper reconciles a NodeKeeper object
+type ReconcileNodeKeeper struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client               client.Client
+	configManagerBuilder configmanager.ConfigManagerBuilder
+	machinery            machinery.Machinery
+	metricsClientBuilder metrics.MetricsBuilder
+	scheme               *runtime.Scheme
+}
+
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNodeKeeper) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+
+	operatorNamespace, err := getOperatorNamespace()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	uc, err := getUpgradeConfigCR(r.client, operatorNamespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	upgradeResult, err := r.machinery.IsUpgrading(r.client, "worker")
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	history := uc.Status.History.GetHistory(uc.Spec.Desired.Version)
+	if !(history.Phase == upgradev1alpha1.UpgradePhaseUpgrading && upgradeResult.IsUpgrading) {
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch the Node instance
+	node := &corev1.Node{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	result := isDraining(node)
+	metricsClient, err := r.metricsClientBuilder.NewClient(r.client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if !result.IsDraining {
```

Comment:
did we need to evaluate this the same as https://github.com/openshift/managed-upgrade-operator/pull/128/files#diff-52b33f64f27b313ab7028c0fa1db2024R105